### ISSUE TITLE
type system design iteration

### ISSUE THREAD

**Issue by JeffBezanson** [JuliaLang/julia#8974]
@jakebolewski, @jiahao and I have been thinking through some type system improvements and things are starting to materialize. I believe this will consist of:
- Use `DataType` to implement tuple types, which will allow for more efficient tuples, and make tuple types less of a special case in various places. This might also open the door for user-defined covariant types in the future. (DONE)
- Add `UnionAll` types around all uses of TypeVars. Each `UnionAll` will bind one TypeVar, and we will nest them to allow for triangular dispatch. So far we've experimented with just sticking TypeVars wherever, and it has not really worked. We don't have a great syntax for this yet, but for now will probably use `@UnionAll T<:Int Array{T}`. These are mostly equivalent to existential types.
- Unspecialized parametric types will actually be UnionAll types. For example `Complex` will be bound to `@UnionAll T<:Real _Complex{T}` where `_Complex` is an internal type constructor.
- Technically, `(Int,String)` could be a subtype of `@UnionAll T (T,T)` if `T` were `Union(Int,String)`, or `Any`. However dispatch doesn't work this way because it's not very useful. We effectively have a rule that if a method parameter `T` only appears in covariant position, it ranges over only concrete types. We should apply this rule explicitly by marking TypeVars as appropriate. So far this is the best way I can think of to model this behavior, but suggestions are sought.
- TypeVars should only be compared by identity, and are not types themselves. We will not have syntax for introducing lone TypeVars; `@UnionAll` should suffice.

This is being prototyped in `examples/juliatypes.jl`. I hope this will fix most of the following issues:

method lookup and sorting issues:
#8959 #8915 #7221 #8163

method ambiguity issues:
#8652 #8045 #6383 #6190 #6187 #6048 #5460 #5384 #3025 #1631 #270

misc. type system:
#8920 #8625 #6984 #6721 #3738 #2552 #8470

In particular, type intersection needs to be trustworthy enough that we can remove the ambiguity warning and generate a runtime error instead.

**Comment by tlycken** [comment]
I just found this thread through a couple of the linked issues, and got a little worried that this might remove one of the IMHO most powerful features of Julia; I really hope I'm misunderstanding the discussion here.

My concern is how `foo(x::T, y::T)` is intended to behave after the changes proposed here. Is the intention that it *should*, or that it *should not*, match `foo(3, 3.0)`? What happens if I have two methods;

```
foo(x::T, y::T) = println("diagonal")
foo(x,y) = println("different")
```

Is the intention that `foo(2, 4.0)` should print `"diagonal"` here, since there is *somewhere* (above `Any`) in the type hierarchy a type `T` such that both `2::T` and `4.0::T`?

**Comment by JeffBezanson** [comment]
No, there is no intent to remove that at all.

**Comment by tlycken** [comment]
Phew =) Thanks for a quick reply!

**Comment by JeffBezanson** [comment]
To expand on that a bit, the goal is

`f{T}(x::T, y::T)` *does not* match `f(2, 4.0)`

`f{T}(x::A{T}, y::T)` *does* match `f(A{Number}(), 1)`

This behavior is basically what we have now, and both cases are intuitive and very useful. The difficulty is in modeling this behavior. It seems clear that in the first case, `T` ranges over only concrete types, and in the second case it ranges over all types.

The first problem is that "all concrete types <: S" denotes the exact same set of values as "S". So the set-theoretic model that works perfectly for the rest of the type system is not as helpful here. Deciding "is S equal to some concrete type?" is not so easy.

The second problem is coming up with notation for this; `UnionAllConcrete` is a bit ugly.

The third problem is that we want to use `UnionAllConcrete` automatically for the first method above, and `UnionAll` for the second one. That's not so hard to do, but when described that way it seems rather ad-hoc. It would be nice to have a more elegant explanation.

**Comment by tlycken** [comment]
Is there another reason for wanting `f{T}(x::A{T}, y::T)` to match `f(A{Number}(), 1)` than notation? In other words, isn't `f{S<:T}(x::A{S}, y::T)` (specifically, `f{T<:Number}(x::Array{T}, y::Number)`) the same thing as the new behavior?

This does make it a little more cumbersome to write your method of `f` matching `(Array{Number}, Int)`, but it's not impossible - and it doesn't introduce the admittedly ad-hoc distinction between when to use the different unions.

**Comment by JeffBezanson** [comment]
Yes, if we had to write it as `f{S,T<:S}(x::A{S}, y::T)` then we could restrict all type variables in covariant position to concrete types, which would be simpler. That seems like a good trade-off.

**Comment by timholy** [comment]
+1 for this suggestion.

**Comment by Jutho** [comment]
Any progress on this? I was thinking of taking a stab at #4774 by implementing a general wrapper
```julia
immutable Transpose{A}
    a::A
end
```
but soon realized that I would need to be able to specify
```
typealias AbstractTransposeMatrix{T} @UnionAll A<:AbstractMatrix{T} Transpose{A}
```
Will these type of constructions be supported?

**Comment by timholy** [comment]
Cross-linking a conversation on julia-dev which referenced this issue as a potential solution to the current performance limitations in constructing types with more than 7 fields:
https://groups.google.com/d/msg/julia-dev/o4BZqLcTorg/sqj7IhR7jFgJ

**Comment by Jutho** [comment]
Thanks for the cross-link. It is cool that this is also covered by this redesign. I hadn't grasped that before.

**Comment by timholy** [comment]
It's still not obvious to me that it is, which is why I inserted the cross-link. Best to make sure it's on the table!

**Comment by mauro3** [comment]
Will this overhaul also fix the performance issues around higher-oder functions (there is no specific issue, I think ) and anonymous functions (#1864)?

**Comment by mauro3** [comment]
How will a parametric types be ordered?  Is there an order, always? Presumably this will this hold as a function with the left signature is more specific:
`@UnionAll T<:Integer (T,T,T) <: @UnionAll T<:Integer (T,T,Integer)`

**Comment by JeffBezanson** [comment]
I really don't think there is time for this in 0.4. It will take a while, and is mostly an algorithmic improvement and new syntax, and not a large breaking change.

**Comment by simonbyrne** [comment]
Unfortunately, I think something like this is more or less essential in order to implement a `Transpose` immutable type, as per #4774.

**Comment by tkelman** [comment]
Could you elaborate on why? I've been tempted to voice a bound-to-be-unpopular opinion that we're running out of time to complete #4774, with enough buffer to make sure it's actually working well, in time for 0.4.0.

**Comment by simonbyrne** [comment]
The current leading contender for #4774 are separate `Transpose` and `Conjugate` types, which would mean that we're going to end up with large hierarchies of nested types. As an example, in the LAPACK code, we will want to dispatch on statements such as "a matrix with `Complex{T<:Union{Float32,Float64}}` entries, its transpose, or its conjugate transpose".

**Comment by timholy** [comment]
Having now read through chunks of Jeff's thesis and spent some time mucking around `jltypes.c`, I'll add a few thoughts (for the 0.5-dev cycle, obviously):
- It's very clear that having a well-founded formal structure, where possible, has always been a major priority. I wonder if we can bump this up yet one more notch, particularly in testing. It seems there are likely to be certain "axioms" that we want our type algorithms to satisfy: I'll propose `typeintersect(A,B) = typeintersect(B,A)` and `is_morespecific(A,B) iff is_morespecific(WrapperType{A}, WrapperType{B})` as two examples. I gather that it may not be possible or straightforward to uniquely specify the behavior of Julia's entire type system in terms of such axioms, but where available such axioms seem like outstanding fodder for building a really comprehensive test suite. (See 4207198683c0732c6402a3b68527425b456c927d for a move in that direction.) The closer we come to spelling out behavior this way, the clearer it may be to users and the easier it might be for not-Jeff developers who, say, spend some time modifying/breaking the type system :smile: and who sometimes wonder what the correct behavior should be (#11242).
- While we have quite a number of tests for intersection in `core.jl` and some tests for subtyping, specificity isn't, as far as I can tell, tested at all. That means we learn about method-sorting problems indirectly, by failures elsewhere, which makes them quite difficult to debug.
- Naively, it seems like it would be awesome if instead of essentially 3 separate implementations of subtyping/specificity (`jl_subtype_le`, `type_morespecific`, and `type_match_`), we could get away with 1 or 2. The current `jl_subtype_le` is not diagonal, and I'm wondering if that should change.

Finally, I have much more intimate experience now with the fact that our current TypeVars are hard to reason about, and I therefore applaud the goal of cleaning this up.

These wishlist items are not idle demands; time permitting (and especially when it comes to testing), I'm happy to help write code. I kind of think we should emphasize writing the tests early in the process, since the test suite will presumably help its development.

**Comment by yuyichao** [comment]
AFAIK, `typeintersect(A, B) == typeintersect(B, A)` is supposed to be true and apparently when it is not the case it worth [mention @JeffBezanson by name](https://github.com/JuliaLang/julia/issues/11136#issuecomment-99491753)

**Comment by timholy** [comment]
Yes (to your edited comment), I thought it was worth testing that basic assumption directly.

That portion of my proposal is essentially stating that I think we should add many more such tests. The risk is that it might result in a `core` test suite that takes a long time to run (particularly if we compose the axioms), but there will be a lot of benefit, too. The type-related algorithms are pretty complicated; in a conceptually-related task (for developing some of the trickier parts of SubArrays, particularly, the linear-indexing dimensionality inference part) I found it essential to test all possible combinations of index types. To me it seems that our type-related test suite is a little on the sparse side. (This doesn't really show up in Coveralls, though, because all that code is implemented in C.)

**Comment by yuyichao** [comment]
I was actually thinking about sth very similar recently.

It would be nice to have some rules to test the implementation against. And maybe also to run combinations of types operations on a set of types to test the validity of the result (although this kind of requires an independent (possibly slower) way to examing the results...).

**Comment by mauro3** [comment]
There was quite a bit of discussion on this over in https://github.com/JuliaLang/julia/issues/11310#issuecomment-170421099

**Comment by timholy** [comment]
When this gets picked up again, one thing worth mentioning: lately, I have been wondering whether a good way to fix many problems with compilation time might be "abstract type inference." The idea would be that when possible, you express the *result* of type inference on a method as a set of `@pure` operations on types. That way, in many cases you'd only have to run type inference once and then you look up the answer each time you want to create a specialization.

For example, in running inference on
```jl
getindex{T,N}(A::Array{T,N}, i) = arrayref(A, i)
```
you might like to express the return value as `::T` "generically" rather than as whatever specific `T`was used for `A`.

In a more complicated case like
```jl
function (+){S,T,N}(::A::Array{S,N}, B::Array{T,N})
    R = promote_op(+, S, T)
    # check sizes...
    out = Array{R}(size(A))
    for I in ...
    end
    out
end
```
then you'd probably like `R` to be annotated `::promote_op(+, S, T)`.

I mention this here because this seems to require the ability to express a type in terms of functions.

**Comment by martinholters** [comment]
Having read through this thread, I'm still not 100% sure of the desired behavior. I got:
* `f{T}(::T, ::T)` should not match `(Int,Float64)`.
* `f{T}(::Array{T}, ::T)` should match `(Array{Number}, Float64)` (with `T==Number`).

But
* should `f{T}(::Array{T}, ::T, ::T)` match `(Array{Number}, Int, Float64)`  (with `T==Number`)?

**Comment by StefanKarpinski** [comment]
If the second matches, I don't see how the third could not match (although I'm not sold on this behavior).

**Comment by timholy** [comment]
I definitely think the second shouldn't match. I like the version in https://github.com/JuliaLang/julia/issues/8974#issuecomment-68657184, and it seems Jeff does too.

**Comment by martinholters** [comment]
Ah, yes, sorry, I had missed https://github.com/JuliaLang/julia/issues/8974#issuecomment-68654723, my bad. Makes a lot more sense that way. Carry on.

**Comment by datnamer** [comment]
So will this fix Union type performance?

**Comment by simonbyrne** [comment]
I came across this recently while using a lot of nested types: I'm not sure if it's something that you're planning on handling or not, but it might be a useful test case:
```
julia> immutable Foo{T} 
       x::T
       end

julia> immutable Bar{T} end

julia> typealias FooBar{T} Foo{Bar{T}}
Foo{Bar{T}}

julia> FooBar{Float64} <: FooBar
false
```

It's been a minor annoyance, as I have to write `f{T}(x::FooBar{T}) = ....` instead of f`(x::FooBar) = ...`.

**Comment by JeffBezanson** [comment]
Yes, the new subtyping algorithm should fix this.

**Comment by JeffBezanson** [comment]
closed by #18457
